NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -1151,3 +1193,76 @@ ULONG CNB::Copy(PVOID Dst, ULONG Length, ULONG DataOffset) const
 
     return Copied;
 }
+
+void CNB::ReturnPages()
+{
+    if (m_ExtraNBStorage)
+    {
+        m_ParentNBL->GetParentTXPath()->ReturnPages(m_ExtraNBStorage);
+        NdisFreeMemory(m_ExtraNBStorage, sizeof(CExtendedNBStorage), 0);
+    }
+    m_ExtraNBStorage = nullptr;
+}
+
+NBMappingStatus CNB::AllocateAndFillCopySGL()
+{
+    //decide how many pages should be used 
+    ULONG uDataLength = GetDataLength();
```

Comment:
Actually we do not need to copy (and to allocate pages for) the starting part of the packet (ParsedHeadersLength bytes). They were already copied to the "headers" page (because the driver in most practical cases need to modify it before sending). It's size is from 0 to entire available area in the "headers" page. In case of TCP packets usually includes IP header + often TCP header.